syntax_and_semantics/capturing_blocks.md

```diff
@@ -37,29 +51,101 @@ model.save # prints "Saved!"
 
 In the above example the type of `&block` wasn't specified: this just means that the captured block doesn't have arguments and doesn't return anything.
 
-Note that if the return type is not specified, nothing gets returned from the proc call:
+## Type restrictions for blocks
+
+There are type restrictions for capturing blocks; there are options beyond requiring one specific input and one specific output, as shown below.
+
+A block can be captured with any of the following type restrictions:
+- Have a specific input and output
+  - `Int32 -> Int32`
+  - `Array(String) -> String`
+- Have any number of inputs (arguments must be a specific type `Int32` not `Int`)
+  - Zero arguments: `->`
+  - One argument: `Int32 ->`
+  - Many arguments: `Int32, String, Array ->`
+- Return a specific type
+  - With no inputs: `-> Int32`, `-> String`
+  - With input(s): `Int32 -> String`
+- Return anything
+  - With no inputs: `-> _`
+  - With input(s): `String -> _`
+- Return nothing (returns `nil`)
+  - With no inputs: `->` (the `->` can be omitted - `def method(&block)` )
+  - With input(s): `Int32 ->`
```

Comment:
This is too verbose. Just saying "Zero or many inputs can be specified, separates by comma. One output can be specified: if it's omitted, it means `Nil`".